src/evaluator.rs

```diff
@@ -1,68 +1,28 @@
-use base64;
 use charset::{self, Charset};
-use quoted_printable;
-use std::{error, fmt, result, str};
 
 use crate::parser::{Ast, Node::*};
 
-#[derive(Debug)]
+#[derive(thiserror::Error, Debug)]
 pub enum Error {
-    DecodeUtf8Error(str::Utf8Error),
-    DecodeBase64Error(base64::DecodeError),
-    DecodeQuotedPrintableError(quoted_printable::QuotedPrintableError),
+    #[error(transparent)]
+    DecodeUtf8Error(#[from] std::str::Utf8Error),
+    #[error(transparent)]
+    DecodeBase64Error(#[from] base64::DecodeError),
+    #[error(transparent)]
+    DecodeQuotedPrintableError(#[from] quoted_printable::QuotedPrintableError),
 }
 
-impl fmt::Display for Error {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match self {
-            Error::DecodeUtf8Error(err) => err.fmt(f),
-            Error::DecodeBase64Error(err) => err.fmt(f),
-            Error::DecodeQuotedPrintableError(err) => err.fmt(f),
-        }
-    }
-}
-
-impl error::Error for Error {
-    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
-        match *self {
-            Error::DecodeUtf8Error(ref err) => Some(err),
-            Error::DecodeBase64Error(ref err) => Some(err),
-            Error::DecodeQuotedPrintableError(ref err) => Some(err),
-        }
-    }
-}
-
-impl From<str::Utf8Error> for Error {
-    fn from(err: str::Utf8Error) -> Error {
-        Error::DecodeUtf8Error(err)
-    }
-}
-
-impl From<base64::DecodeError> for Error {
-    fn from(err: base64::DecodeError) -> Error {
-        Error::DecodeBase64Error(err)
-    }
-}
-
-impl From<quoted_printable::QuotedPrintableError> for Error {
-    fn from(err: quoted_printable::QuotedPrintableError) -> Error {
-        Error::DecodeQuotedPrintableError(err)
-    }
-}
-
-type Result<T> = result::Result<T, Error>;
-
-fn decode_utf8(encoded_bytes: &Vec<u8>) -> Result<&str> {
-    let decoded_bytes = str::from_utf8(&encoded_bytes)?;
+fn decode_utf8(encoded_bytes: &Vec<u8>) -> Result<&str, Error> {
+    let decoded_bytes = std::str::from_utf8(&encoded_bytes)?;
     Ok(decoded_bytes)
 }
 
-fn decode_base64(encoded_bytes: &Vec<u8>) -> Result<Vec<u8>> {
+fn decode_base64(encoded_bytes: &Vec<u8>) -> Result<Vec<u8>, Error> {
```

Comment:
What about doing an alias next to the `Error` definition? `type Result<T> = std::result::Result<T, Error>`.

Reply:
I don't really see any advantages in this... but also no disadvantages, so yeah. I just fixed that :)

Reply:
I guess the only advantage is not to repeat the error each time, you save some keystrokes :smile:

Reply:
Welp, programmer type-laziness at its finest :rofl: